doc/man/ustctl.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "USTCTL" "1" "January 2011" "" ""
.
.SH "NAME"
\fBustctl\fR \- a program to control the tracing of userspace applications
.
.SH "SYNOPSIS"
\fBustctl\fR [\fICOMMAND\fR] [\fIARGS\fR]\.\.\.
.
.SH "DESCRIPTION"
\fBustctl\fR is a program to control the tracing of userspace applications\. It can list markers, start the tracing, stop the tracing, enable/disable markers, etc\.
.
.SH "OPTIONS"
These programs follow the usual GNU command line syntax, with long options starting with two dashes(`\-\')\. A summary of options is included below\.
.
.TP
\fB\-h\fR, \fB\-\-help\fR
Show summary of commands\.
.
.SH "COMMANDS"
\fBustctl\fR accepts commands followed by arguments for each respective command\. Most commands require the pid of the application being traced\.
.
.TP
\fBcreate\-trace\fR \fIPID\fR \fITRACE\fR
Create trace\.
.
.TP
\fBalloc\-trace\fR \fIPID\fR \fITRACE\fR
Allocate trace\.
.
.TP
\fBstart\-trace\fR \fIPID\fR \fITRACE\fR
Start tracing\.
.
.TP
\fBstop\-trace\fR \fIPID\fR \fITRACE\fR
Stop tracing\.
.
.TP
\fBdestroy\-trace\fR \fIPID\fR \fITRACE\fR
Destroy the trace\.
.
.TP
\fBset\-subbuf\-size\fR \fIPID\fR \fITRACE\fR \fICHANNEL\fR/\fIbytes\fR
Set the size of subbuffers in CHANNEL\.
.
.TP
\fBset\-subbuf\-num\fR \fIPID\fR \fITRACE\fR \fICHANNEL\fR/\fInr\fR
Set the number of subbuffers per buffer for CHANNEL\. Must be a power of 2\.
.
.TP
\fBset\-sock\-path\fR \fIPID\fR \fISOCKPATH\fR
Set the path of the daemon socket\.
.
.TP
\fBget\-subbuf\-size\fR \fIPID\fR \fITRACE\fR \fICHANNEL\fR
Print the size of subbuffers per buffer for CHANNEL\.
.
.TP
\fBget\-subbuf\-num\fR \fIPID\fR \fITRACE\fR \fICHANNEL\fR
Print the number of subbuffers per buffer for CHANNEL\.
.
.TP
\fBget\-sock\-path\fR \fIPID\fR
Get the path of the daemon socket\.
.
.TP
\fBenable\-marker\fR \fIPID\fR \fITRACE\fR \fICHANNEL\fR/\fIMARKER\fR
Enable a marker\.
.
.TP
\fBdisable\-marker\fR \fIPID\fR \fITRACE\fR \fICHANNEL\fR/\fIMARKER\fR
Disable a marker\.
.
.TP
\fBlist\-markers\fR \fIPID\fR
List the markers of the process, their state and format string\.
.
.TP
\fBforce\-subbuf\-switch\fR \fIPID\fR \fITRACE\fR
Force a subbuffer switch\. This will flush all the data currently held\.
.
.SH "LIFE CYCLE OF A TRACE"
Typically, the first step is to enable markers with \fBenable\-marker\fR\. An enabled marker generates an event when the control flow passes over it (assuming the trace is recording)\. A disabled marker produces nothing\. Enabling and disabling markers may however be done at any point, including while the trace is being recorded\.
.
.P
In order to record events, a trace is first created with \fBcreate\-trace\fR\. At this point, the subbuffer count and size may be changed with \fBset\-subbuf\-num\fR and \fBset\-subbuf\-size\fR\.
.
.P
Afterward, the trace may be allocated with \fBalloc\-trace\fR\. This allocates the buffers in memory, so once this is done, the subbuffer size and count can not be changed\. Trace allocation also causes the daemon to connect to the trace buffers and wait for data to arrive\. Explicit allocation is optional, as it is done automatically at trace start\.
.
.P
The trace may then be started with \fBstart\-trace\fR\. This results in events being recorded in the buffer\. The daemon automatically collects these events\.
.
.P
The trace may be stopped with \fBstop\-trace\fR, either definitely after all the wanted information is collected, or temporarily, before being started again with \fBstart\-trace\fR\. This results in effectively \'pausing\' the recording\. After using \fBstop\-trace\fR, if a daemon is collecting the trace, it may not have flushed to the disk the full contents of the buffer yet\.
.
.P
Finally, when \fBdestroy\-trace\fR is used, the trace buffers are unallocated\. However, the memory may not be effectively freed until the daemon finishes to collect them\. When the trace is being collected by \fBust\-consumerd\fR, this command guarantees its full contents is flushed to the disk\.
.
.SH "STRUCTURE OF A TRACE"
Each instrumentation point that is added in a program is associated to a channel\.
.
.P
Trace events are put in buffers\. There is one buffer per channel, per cpu\. For example, on a system with 4 cores and tracing an application with 3 channels, there will be 12 buffers in total\. The content of each of these buffers is put in a distinct file in the trace directory\. For example, the \fBmetadata_2\fR file contains the data that was extracted from the buffer that contained the events from the metadata channel and having occurred on cpu 2\.
.
.P
In memory, each buffer is divided in subbuffers\. Subbuffers are equally\-sized, contiguous parts of a buffer\. The size of a buffer is equal to the number of subbuffers it contains times the size of each subbuffer\. When a subbuffer is full, it is collected by the daemon while the others are filled\. If, however, the buffer size is too small, buffer overflows may occur and result in event loss\. By default, the number of subbuffers per buffer is 2\. Subbuffer size for a given channel may be chosen with \fBset\-subbuf\-size\fR while the subbuffer count is set with \fBset\-subbuf\-num\fR\.
.
.SH "SEE ALSO"
usttrace(1), ust\-consumerd(1)
.
.SH "AUTHOR"
\fBustctl\fR was written by Pierre\-Marc Fournier\.
.
.P
This manual page was written by Jon Bernard <jbernard@debian\.org>, for the Debian project (and may be used by others)\. It was updated by Pierre\-Marc Fournier\.


ChangeLog
2011-01-14 ust 0.11
	* Fix a serious bug in the newly introduced synchronized timer
	implementation.
	* Increase compatibility of an autoconf test

2011-01-14 ust 0.10
	* Run time check for lttng clock source providing synchronized
	timestamps
	* Rename of ustd and libustd to ust-consumerd and libustconsumer,
	remember to run ldconfig

2010-12-03 ust 0.9
	* Cleaned up namespace
	* Compatible with userspace-rcu version 0.5.2
	* Fixed a minor bug in a test

2010-10-26 ust 0.8
	* API CHANGE!!! Add data pointer to tracepoint probe
	void *data is always the first argument to a tracepoint probe
	* License LGPGv2.1 text and information added
	* Update COPYING file
	* Add TRACE_EVENT api proposal, not finalised
	NOT usable at this point. Only bound to a printf.
	ustctl command for listing trace event
	* Multiple fixes
	Compilation warnings
	Value and error handling
    	Fix bugs introduce by prior commits
	Security fix (CVE-2010-3386)
	Add exit status for ustctl
	Change malloc to zmalloc for most of UST
	Remove unused code, syntax cleanup and consistency for markers
	* New UST internal communication protocol
	Add message headers
	Usage of epoll
	* New testcases
	* Try to decrease namespace pollution
	* Fix valgrind for sendmsg bug
	* tracectl destructor aware of fork

2010-08-16  ust 0.6
	* add ust strncpy interface for input robustness
	* multiple memory barrier fix
	* add libustd: contains the core functionality of ustd
	* add get/set commands for daemon socket path
	* fix cacheline aligned to use liburcu value
	* fix usage of custom probes: use trace_mark_tp
	* remove ltt_relay_buffer_flush function
	* add ustctl manpage: force-switch and get/set-sock-path
	* fix invalid array size and directory creation mode for ustd

2010-06-30  ust 0.5
	* add support for PowerPC 32 architecture
	* use clock_gettime(CLOCK_MONOTINIC) instead of gettimeofday()
	* add several tests
	* output test suite results in TAP format
	* add support for traced library unregistration (useful with dlclose())
	* add manpages
	* compile with -fno-strict-antialiasing
	* add support for channel overwrite and non-collection
	* update tracepoint api from kernel
	* fix various bugs and do various cleanups
	* improve documentation

2010-03-17  ust 0.4
	* remove dependency on libkcompat
	* use atomic ops from liburcu
	* libmallocwrap: rename to libustinstr-malloc
	* libinterfork: rename to libustfork
	* add a testsuite
	* add ustctl --get-subbuf-size and --get-subbuf-num
	* updates for compilation with gcc-4.4
	* remove periodical wakeup in listener thread
	* change benign warnings to debug messages, not to confuse users
	* ustd: fix destruction to free all file descriptors and memory
	* libust: close all file descriptors on close
	* remove periodical wakeup in daemon
	* code cleanups

ust 0.3
	* fix several issues, stabilizing tracing of large applications
	* add --set-subbuf-size / --set-subbuf-num options to ustctl
	* add a signal safe implementation of snprintf

ust 0.2
	* conversion of the buffering system to per-cpu
	* update to the liburcu 0.4 API
	* upgrade to trace format 2.6; the latest lttv may now be used
	* add support for clone() interception in libinterfork
	* port refactorings from the kernel tracer up to 0.173
	* enhancements/refactor of libustcomm
	* add sections in the manual about trace viewing and about tracing apps not linked to libust
	* debugging output is now enabled by adding -DUST_DEBUG to the CFLAGS
	* usttrace: change ustd socket name to ustd-sock-PID
	* change app socket directory name for something more descriptive (/tmp/ust-app-socks)
	* code cleanups
	* various bug fixes


ustctl/ustctl.c
/* Copyright (C) 2009  Pierre-Marc Fournier
 * Copyright (C) 2011  Ericsson AB, Nils Carlson <nils.carlson@ericsson.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>
#include <fcntl.h>

#include "ust/ustctl.h"
#include "usterr.h"
#include "cli.h"
#include "scanning_functions.h"

void usage(const char *process_name)
{
	fprintf(stderr, "Usage: %s COMMAND [ARGS]...\n", process_name);
	fprintf(stderr,
		"Control tracing within a process that supports UST,\n"
		" the Userspace Tracing libary\n"
		"Options:\n"
		"  -h[<cmd>], --help[=<cmd>]        "
		"help, for a command if provided\n"
		"  -l, --list                       "
		"short list of commands\n"
		"  -e, --extended-list              "
	       "extented list of commands with help\n"
		"Commands:\n");
	list_cli_cmds(CLI_DESCRIPTIVE_LIST);
}

struct option options[] =
{
	{"help", 2, NULL, 'h'},
	{"list", 0, NULL, 'l'},
	{"extended-list", 0, NULL, 'e'},
	{NULL, 0, NULL, 0},
};

int main(int argc, char *argv[])
{
	struct cli_cmd *cli_cmd;
	int opt;

	if(argc <= 1) {
		fprintf(stderr, "No operation specified.\n");
		usage(argv[0]);
		exit(EXIT_FAILURE);
	}

	while ((opt = getopt_long(argc, argv, "+h::le",
				  options, NULL)) != -1) {
		switch (opt) {
		case 'h':
			if (!optarg) {
				usage(argv[0]);
			} else {
				if (cli_print_help(optarg)) {
					fprintf(stderr, "No such command %s\n",
						optarg);
				}
			}
			exit(EXIT_FAILURE);
			break;
		case 'l':
			list_cli_cmds(CLI_SIMPLE_LIST);
			exit(EXIT_FAILURE);
			break;
		case 'e':
			list_cli_cmds(CLI_EXTENDED_LIST);
			exit(EXIT_FAILURE);
		default:
			fprintf(stderr, "Unknown option\n");
			break;
		}
	}

	cli_cmd = find_cli_cmd(argv[optind]);
	if (!cli_cmd) {
		fprintf(stderr, "No such command %s\n",
			argv[optind]);
		exit(EXIT_FAILURE);
	}

	cli_dispatch_cmd(cli_cmd, argc - optind, &argv[optind]);

	return 0;
}

static int list_trace_events(int argc, char *argv[])
{
	struct trace_event_status *tes = NULL;
	int i, sock;

	sock = parse_and_connect_pid(argv[1]);

	if (ustctl_get_tes(sock, &tes)) {
		ERR("error while trying to list "
		    "trace_events for PID %s\n",
		    argv[1]);
		return -1;
	}
	i = 0;
	for (i = 0; tes[i].name; i++) {
		printf("{PID: %s, trace_event: %s}\n",
		       argv[1],
		       tes[i].name);
	}
	ustctl_free_tes(tes);

	return 0;
}

static int set_sock_path(int argc, char *argv[])
{
	int sock;

	sock = parse_and_connect_pid(argv[1]);

	if (ustctl_set_sock_path(sock, argv[2])) {
		ERR("error while trying to set sock path for PID %s\n", argv[1]);
		return -1;
	}

	return 0;
}

static int get_sock_path(int argc, char *argv[])
{
	int sock;
	char *sock_path;

	sock = parse_and_connect_pid(argv[1]);

	if (ustctl_get_sock_path(sock, &sock_path)) {
		ERR("error while trying to get sock path for PID %s\n", argv[1]);
		return -1;
	}
	printf("The socket path is %s\n", sock_path);
	free(sock_path);

	return 0;
}

struct cli_cmd __cli_cmds general_cmds[] = {
	{
		.name = "list-trace-events",
		.description = "List trace-events for a given pid",
		.help_text = "list-trace-events <pid>\n"
		"List the trace-events in a process\n",
		.function = list_trace_events,
		.desired_args = 1,
		.desired_args_op = CLI_EQ,
	},
	{
		.name = "set-sock-path",
		.description = "Set the path to the consumer daemon socket",
		.help_text = "set-sock-path <pid> <sock-path>\n"
		"Set the path to the consumer daemon socket\n",
		.function = set_sock_path,
		.desired_args = 2,
		.desired_args_op = CLI_EQ,
	},
	{
		.name = "get-sock-path",
		.description = "Get the path to the consumer daemon socket",
		.help_text = "get-sock-path <pid>\n"
		"Get the path to the consumer daemon socket\n",
		.function = get_sock_path,
		.desired_args = 1,
		.desired_args_op = CLI_EQ,
	},
};
